### DISCUSSION TITLE
Cannot send `Request` to web worker

### DISCUSSION THREAD

**Discussion by rijenkii** [denoland/deno#17437] [Q&A] [answered]
Source:
```typescript
// main.ts
import { serve } from "https://deno.land/std@0.172.0/http/server.ts";

const worker = new Worker(new URL("./worker.ts", import.meta.url), { type: "module" });
serve((req) => {
  worker.postMessage({ id: Math.random(), req });
  return new Response("hello, world!");
});
```

```typescript
// worker.ts
self.onmessage = (ev: MessageEvent<{ id: number, req: Request }>) => console.log(ev.data);
```

Output:
```
$ deno run --allow-read --allow-net main.ts &
Listening on http://localhost:8000/
$ curl -s localhost:8000 >/dev/null
{ id: 0.6638532911173409, req: {} }
```

`Request` object is replaced with `{}` in worker.

Question: why? Is it forbidden? Should I deconstruct the request before sending it?

**Comment by augustollenz** [comment, marked as answer]
Note that `postMessage` uses `structuredClone algorithm` [1] and `Request` type is not listed as a supported type [2] to be serialized by that algorithm. I don't really know the details about this subject, but I suppose that this is why your code didn't work.

[1] https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#parameters
[2] https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#webapi_types

**Reply by andreubotella** [reply]
That is the answer, but also, Deno doesn't currently support structured cloning of any web APIs. The issue is #12067, and adding support for this turned out to be much trickier than expected because of the complexity of adding a new `rusty_v8` scope (denoland/rusty_v8#862).
